### PR TITLE
http protocol detector

### PR DESCRIPTION
**Pull request by suncairong163**
## What is the purpose of the change

add http protocol detector for support  http PortUnification

## Verifying this change

@AlbumenJ 

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by suncairong163** [comment]
@chickenlj    rest protocol  is integrated into   PortUnification

**Review by AlbumenJ — APPROVED** [review]

**Comment by AlbumenJ** [comment]
@EarthChen @CrazyHZM PTAL

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/HttpUtils.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import static org.apache.dubbo.common.utils.HttpUtils.HttpMethod.PUT;
+import static org.apache.dubbo.common.utils.HttpUtils.HttpMethod.HEAD;
+import static org.apache.dubbo.common.utils.HttpUtils.HttpMethod.PATCH;
+import static org.apache.dubbo.common.utils.HttpUtils.HttpMethod.OPTIONS;
+import static org.apache.dubbo.common.utils.HttpUtils.HttpMethod.TRACE;
+
+/**
+ * for http methods
+ */
+public class HttpUtils {
```
Comment (CrazyHZM) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/HttpUtils.java:35]: This class is not suitable for the dubco-common module in terms of function and methods. There is no current intersection between qos and http handling except HttpMethod, and the methods in HttpUtils are not suitable for all modules.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/HttpUtils.java (hunk 1 above)
Comment (suncairong163) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/HttpUtils.java:35]: ok , I will change the position

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/ProtocolDetector.java [hunk 2]
```diff
@@ -17,18 +17,90 @@
...
- *
  */
 public interface ProtocolDetector {
+    int empty = ' ';
+
+    default Result detect(ChannelBuffer in) {
+        return Result.UNRECOGNIZED;
+    }
+
+    default Result detect(ChannelBuffer in, URL url) {
```
[the remark this answers is not in the document]
Comment (suncairong163) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/ProtocolDetector.java:34]: 1.qos和rest 在http检测这块是有冲突的， 当Qos 和rest同时开启时， 需要进行 requestUrl 判断 ， service经过 RestProtocol 进行export时才认为 rest开启，不能通过RestProtocol的存在与否判断，这个rest开启标志通过url进行传递
 2.希望detector添加 开关，通过url属性进行判断
 所以就增加了这个参数

[the remark this answers is not in the document]
Comment (CrazyHZM) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/ProtocolDetector.java:34]: 1. 我在想qos有关http的是不是还是需要依赖http的模块，这样才能保证所有http的实现都在一个地方，它的层级应该跟triple是一样的，应该在抽象的http之上，qos的命令应该只是rest服务接口中的一部分，但是qos命令受开关影响，举个例子，有一个AbstractHttpProtocolDetector。
2. 这个判断应该是有export的时候就应该把开关传递过来，就像QosEnableFlag，但我觉得这个开关的确可以被抽象在ProtocolDetector中。

[the remark this answers is not in the document]
Comment (suncairong163) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/ProtocolDetector.java:34]: 1.AbstractHttpProtocolDetector 这个抽象可以的有qos和http两个实现，ProtocolDetector 直接进行enable的set,因为QosWireProtocol 这个是单例的，不同url之间开关会不会有影响？增加url参数也是想动态的进行开关，还有就是如果用户想显式的关闭某种协议就不能单单通过export判断了，
 
 2.NettyPortUnificationServer 这个server实现上，只会保存第一个实例创建时的URL,而且后面的执行链路，相同端口使用的都是这个url,不同的service的上下文就丢掉了
 3.目前的单端口多协议wireProtocol 是分开实现的，如果用户想实现所有协议支持，就需要引入dubboall 或者每个rpc protcol 都要引入进入

[the remark this answers is not in the document]
Comment (suncairong163) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/ProtocolDetector.java:34]: AbstractHttpProtocolDetector 这个抽象可以的有qos和http两个实现，ProtocolDetector 直接进行enable的set,因为QosWireProtocol 这个是单例的，不同url之间开关会不会有影响？增加url参数也是想动态的进行开关，还有就是如果用户想显式的关闭某种协议就不能单单通过export判断了

**Review by CrazyHZM — APPROVED** [review]
LGTM

**Pull request merged**